lib/ansible/parsing/metadata.py

```diff
@@ -0,0 +1,208 @@
+__metaclass__ = type
+
+import ast
+
+import yaml
+
+from ansible.module_utils._text import to_text
+
+
+class ParseError(Exception):
+    """Thrown when parsing a file fails"""
+    pass
+
+
+def seek_end_of_dict(module_data, start_line, start_col, next_node_line, next_node_col):
+    """Look for the end of a dict in a set of lines
+
+    We know the starting position of the dict and we know the start of the
+    next code node but in between there may be multiple newlines and comments.
+    There may also be multiple python statements on the same line (separated
+    by semicolons)
+
+    Examples::
+        ANSIBLE_METADATA = {[..]}
+        DOCUMENTATION = [..]
+
+        ANSIBLE_METADATA = {[..]} # Optional comments with confusing junk => {}
+        # Optional comments {}
+        DOCUMENTATION = [..]
+
+        ANSIBLE_METADATA = {
+            [..]
+            }
+        # Optional comments {}
+        DOCUMENTATION = [..]
+
+        ANSIBLE_METADATA = {[..]} ; DOCUMENTATION = [..]
+
+        ANSIBLE_METADATA = {}EOF
+    """
+    if next_node_line is None:
+        # The dict is the last statement in the file
+        snippet = module_data.splitlines()[start_line:]
+        next_node_col = 0
+        # Include the last line in the file
+        last_line_offset = 0
+    else:
+        # It's somewhere in the middle so we need to separate it from the rest
+        snippet = module_data.splitlines()[start_line:next_node_line]
+        # Do not include the last line because that's where the next node
+        # starts
+        last_line_offset = 1
+
+    if next_node_col == 0:
+        # This handles all variants where there are only comments and blank
+        # lines between the dict and the next code node
+
+        # Step backwards through all the lines in the snippet
+        for line_idx, line in tuple(reversed(tuple(enumerate(snippet))))[last_line_offset:]:
+            end_col = None
+            # Step backwards through all the characters in the line
+            for col_idx, char in reversed(tuple(enumerate(c for c in line))):
+                if char == '}' and end_col is None:
+                    # Potentially found the end of the dict
+                    end_col = col_idx
+
+                elif char == '#' and end_col is not None:
+                    # The previous '}' was part of a comment.  Keep trying
+                    end_col = None
+
+            if end_col is not None:
+                # Found the end!
+                end_line = start_line + line_idx
+                break
+    else:
+        # Harder cases involving multiple statements on one line
+        # Good Ansible Module style doesn't do this so we're just going to
+        # treat this as an error for now:
+        raise ParseError('Multiple statements per line confuses the module metadata parser.')
+
+    return end_line, end_col
+
+
+def seek_end_of_string(module_data, start_line, start_col, next_node_line, next_node_col):
```

Comment:
This method is called with only one argument, but there are no defaults given for the unused arguments.

If it is called, the `NotImplementedError` will never be reached.